engine/src/flutter/shell/platform/windows/host_window_dialog.cc

```diff
@@ -0,0 +1,138 @@
+  if (owner_window) {
+    // If the owner window has WS_EX_TOOLWINDOW style, apply the same
+    // style to the dialog.
+    if (GetWindowLongPtr(owner_window, GWL_EXSTYLE) & WS_EX_TOOLWINDOW) {
+      extended_window_style |= WS_EX_TOOLWINDOW;
+    }
+  }
+  return extended_window_style;
+}
+}  // namespace
+
+namespace flutter {
+
+HostWindowDialog::HostWindowDialog(WindowManager* window_manager,
+                                   FlutterWindowsEngine* engine,
+                                   const WindowSizeRequest& preferred_size,
+                                   const BoxConstraints& constraints,
+                                   LPCWSTR title,
+                                   HWND owner_window)
+    : HostWindow(
+          window_manager,
+          engine,
+          WindowArchetype::kDialog,
+          GetWindowStyleForDialog(owner_window),
+          GetExtendedWindowStyleForDialog(owner_window),
+          constraints,
+          [&]() -> Rect {
+            auto const window_style = GetWindowStyleForDialog(owner_window);
+            auto const extended_window_style =
+                GetExtendedWindowStyleForDialog(owner_window);
+            std::optional<Size> const window_size = GetWindowSizeForClientSize(
+                *engine->windows_proc_table(),
+                Size(preferred_size.preferred_view_width,
+                     preferred_size.preferred_view_height),
+                constraints.smallest(), constraints.biggest(), window_style,
+                extended_window_style, owner_window);
+            Point window_origin = {CW_USEDEFAULT, CW_USEDEFAULT};
+            if (owner_window != nullptr && window_size.has_value()) {
+              // Center dialog in the owner's frame.
+              RECT frame;
+              DwmGetWindowAttribute(owner_window, DWMWA_EXTENDED_FRAME_BOUNDS,
+                                    &frame, sizeof(frame));
+              window_origin = {
+                  (frame.left + frame.right - window_size->width()) * 0.5,
+                  (frame.top + frame.bottom - window_size->height()) * 0.5};
+            }
+
+            return {window_origin, window_size
+                                       ? *window_size
+                                       : Size{CW_USEDEFAULT, CW_USEDEFAULT}};
+          }(),
+          title,
+          owner_window) {
+  auto hwnd = window_handle_;
+  if (owner_window == nullptr) {
+    if (HMENU hMenu = GetSystemMenu(hwnd, FALSE)) {
+      EnableMenuItem(hMenu, SC_CLOSE, MF_BYCOMMAND | MF_GRAYED);
+    }
+  }
+
+  if (owner_window != nullptr) {
+    UpdateModalState();
+  }
+}
+
+LRESULT HostWindowDialog::HandleMessage(HWND hwnd,
+                                        UINT message,
+                                        WPARAM wparam,
+                                        LPARAM lparam) {
+  switch (message) {
+    case WM_DESTROY:
+      is_being_destroyed_ = true;
+      if (HostWindow* const owner_window = GetOwnerWindow()) {
```

Comment:
`WM_DESTROY`'s [docs](https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-destroy) says:

> This message is sent first to the window being destroyed and then to the child windows (if any) as they are destroyed. During the processing of the message, it can be assumed that all child windows still exist.

Just to confirm: if the owner window was destroyed (thereby causing this modal dialog window to also get destroyed), `GetOwnerWindow()` will return `nullptr`?